File: src/memory/memory-system.ts
/**
 * Memory System for AI Trading Agent
 * Provides short-term and long-term memory to maintain continuity
 */

import type { IDatabase, TradeRecord, PerformanceMetrics, MarketSnapshot } from '../types/index.js';

export interface ShortTermMemory {
  recentTrades: TradeRecord[];           // Last N trades
  recentDecisions: string[];             // Last N decision reasonings
  currentStreak: {
    type: 'winning' | 'losing';
    count: number;
  };
  recentMarketConditions: string[];      // Description of recent market states
}

export interface LongTermMemory {
  keyLessons: Lesson[];                  // Important learnings
  successPatterns: Pattern[];            // Winning trade patterns
  failurePatterns: Pattern[];            // Losing trade patterns
  marketRegimes: MarketRegime[];         // Different market conditions experienced
  personalityTraits: PersonalityTraits;  // Evolved trading personality
}

export interface Lesson {
  id?: number;
  timestamp: number;
  category: 'success' | 'failure' | 'insight';
  content: string;                       // The lesson learned
  context: string;                       // Market conditions when learned
  importance: number;                    // 1-10 importance score
  applicationCount: number;              // How many times applied
}

export interface Pattern {
  id?: number;
  description: string;
  indicators: Record<string, any>;       // Technical conditions
  outcome: 'positive' | 'negative';
  frequency: number;
  avgReturn: number;
  confidence: number;
}

export interface MarketRegime {
  id?: number;
  name: string;                          // e.g., "High Volatility Bull", "Low Vol Bear"
  characteristics: string;
  performance: number;                   // How well we performed
  tradeCount: number;
  lastSeen: number;
}

export interface PersonalityTraits {
  riskTolerance: number;                 // 0-1, evolved from performance
  aggressiveness: number;                // 0-1, how quickly to enter trades
  patience: number;                      // 0-1, willingness to wait for setups
  adaptability: number;                  // 0-1, how quickly to change strategy
  confidenceLevel: number;               // 0-1, based on recent performance
  tradingPhilosophy: string;             // Evolved description of approach
}

export class MemorySystem {
  constructor(private database: IDatabase) {}

  /**
   * Build short-term memory from recent activity
   */
  async buildShortTermMemory(limit: number = 10): Promise<ShortTermMemory> {
    // Get recent trades
    const recentTrades = await this.database.getTrades({
      status: 'closed',
      limit,
    });

    // Get recent decisions (last 20)
    const recentDecisions = await this.database.getDecisions({ limit: 20 });
    const reasonings = recentDecisions
      .filter(d => d.riskApproved)
      .map(d => {
        try {
          const decision = JSON.parse(d.aiOutput);
          return decision.reasoning || '';
        } catch {
          return '';
        }
      })
      .filter(Boolean)
      .slice(0, limit);

    // Calculate current streak
    const streak = this.calculateStreak(recentTrades);

    // Describe recent market conditions
    const marketConditions = this.describeRecentMarket(recentTrades);

    return {
      recentTrades,
      recentDecisions: reasonings,
      currentStreak: streak,
      recentMarketConditions: marketConditions,
    };
  }

  /**
   * Build long-term memory from historical data
   */
  async buildLongTermMemory(): Promise<LongTermMemory> {
    // Get all closed trades for pattern analysis
    const allTrades = await this.database.getTrades({ status: 'closed' });

    // Extract key lessons
    const keyLessons = await this.extractKeyLessons(allTrades);

    // Identify success and failure patterns
    const successPatterns = await this.identifyPatterns(allTrades, 'positive');
    const failurePatterns = await this.identifyPatterns(allTrades, 'negative');

    // Identify market regimes
    const marketRegimes = await this.identifyMarketRegimes(allTrades);

    // Build personality traits based on trading history
    const personalityTraits = await this.buildPersonalityTraits(allTrades);

    return {
      keyLessons,
      successPatterns,
      failurePatterns,
      marketRegimes,
      personalityTraits,
    };
  }

  /**
   * Calculate current winning/losing streak
   */
  private calculateStreak(trades: TradeRecord[]): { type: 'winning' | 'losing'; count: number } {
    if (trades.length === 0) {
      return { type: 'winning', count: 0 };
    }

    // Sort by timestamp descending (most recent first)
    const sorted = [...trades].sort((a, b) => b.timestamp - a.timestamp);

    let count = 0;
    const firstTradePnl = sorted[0].pnl || 0;
    const type: 'winning' | 'losing' = firstTradePnl > 0 ? 'winning' : 'losing';

    for (const trade of sorted) {
      const pnl = trade.pnl || 0;
      if ((type === 'winning' && pnl > 0) || (type === 'losing' && pnl < 0)) {
        count++;
      } else {
        break;
      }
    }

    return { type, count };
  }

  /**
   * Describe recent market conditions
   */
  private describeRecentMarket(trades: TradeRecord[]): string[] {
    if (trades.length === 0) {
      return ['No recent trading history'];
    }

    const descriptions: string[] = [];

    // Analyze volatility from recent trades
    const recentPnlPercents = trades.slice(0, 5).map(t => Math.abs(t.pnlPercent || 0));
    const avgVolatility = recentPnlPercents.reduce((sum, v) => sum + v, 0) / recentPnlPercents.length;

    if (avgVolatility > 10) {
      descriptions.push('High volatility environment with large price swings');
    } else if (avgVolatility > 5) {
      descriptions.push('Moderate volatility with normal price movements');
    } else {
      descriptions.push('Low volatility, ranging market conditions');
    }

    // Analyze win rate trend
    const recentWins = trades.slice(0, 10).filter(t => (t.pnl || 0) > 0).length;
    const recentWinRate = (recentWins / Math.min(10, trades.length)) * 100;

    if (recentWinRate > 60) {
      descriptions.push('Recent trades have been mostly profitable');
    } else if (recentWinRate < 40) {
      descriptions.push('Struggling with recent trades, need to reassess strategy');
    }

    return descriptions;
  }

  /**
   * Extract key lessons from trading history
   */
  private async extractKeyLessons(trades: TradeRecord[]): Promise<Lesson[]> {
    const lessons: Lesson[] = [];

    if (trades.length === 0) {
      return lessons;
    }

    // Lesson 1: Best performing trade type
    const longTrades = trades.filter(t => t.action === 'OPEN_LONG');
    const shortTrades = trades.filter(t => t.action === 'OPEN_SHORT');

    const longPnl = longTrades.reduce((sum, t) => sum + (t.pnl || 0), 0);
    const shortPnl = shortTrades.reduce((sum, t) => sum + (t.pnl || 0), 0);

    if (longTrades.length > 0 && shortTrades.length > 0) {
      if (longPnl > shortPnl * 1.5) {
        lessons.push({
          timestamp: Date.now(),
          category: 'success',
          content: 'Long positions have been significantly more profitable than short positions. Focus on bullish setups.',
          context: `${longTrades.length} longs vs ${shortTrades.length} shorts`,
          importance: 8,
          applicationCount: 0,
        });
      } else if (shortPnl > longPnl * 1.5) {
        lessons.push({
          timestamp: Date.now(),
          category: 'success',
          content: 'Short positions have been more profitable. Look for bearish setups and resistance levels.',
          context: `${shortTrades.length} shorts vs ${longTrades.length} longs`,
          importance: 8,
          applicationCount: 0,
        });
      }
    }

    // Lesson 2: Optimal leverage
    const leveragePerformance = new Map<number, { pnl: number; count: number }>();
    for (const trade of trades) {
      const current = leveragePerformance.get(trade.leverage) || { pnl: 0, count: 0 };
      current.pnl += trade.pnl || 0;
      current.count++;
      leveragePerformance.set(trade.leverage, current);
    }

    let bestLeverage = 1;
    let bestAvgPnl = -Infinity;
    leveragePerformance.forEach((data, leverage) => {
      const avgPnl = data.pnl / data.count;
      if (avgPnl > bestAvgPnl && data.count >= 3) {
        bestAvgPnl = avgPnl;
        bestLeverage = leverage;
      }
    });

    if (bestLeverage > 0) {
      lessons.push({
        timestamp: Date.now(),
        category: 'insight',
        content: `${bestLeverage}x leverage has yielded the best risk-adjusted returns. Consider using this as default.`,
        context: `Analyzed ${trades.length} trades across different leverage levels`,
        importance: 7,
        applicationCount: 0,
      });
    }

    // Lesson 3: Biggest losses (failure lessons)
    const biggestLosses = trades
      .filter(t => (t.pnl || 0) < 0)
      .sort((a, b) => (a.pnl || 0) - (b.pnl || 0))
      .slice(0, 3);

    for (const loss of biggestLosses) {
      lessons.push({
        timestamp: loss.timestamp,
        category: 'failure',
        content: `Large loss on ${loss.symbol}: ${loss.reasoning}. Exit plan was: ${loss.exitPlan.invalidation}`,
        context: `Lost $${Math.abs(loss.pnl || 0).toFixed(2)} (${Math.abs(loss.pnlPercent || 0).toFixed(2)}%)`,
        importance: 9,
        applicationCount: 0,
      });
    }

    // Lesson 4: Best wins (success lessons)
    const biggestWins = trades
      .filter(t => (t.pnl || 0) > 0)
      .sort((a, b) => (b.pnl || 0) - (a.pnl || 0))
      .slice(0, 3);

    for (const win of biggestWins) {
      lessons.push({
        timestamp: win.timestamp,
        category: 'success',
        content: `Successful trade on ${win.symbol}: ${win.reasoning}`,
        context: `Earned $${(win.pnl || 0).toFixed(2)} (${(win.pnlPercent || 0).toFixed(2)}%)`,
        importance: 8,
        applicationCount: 0,
      });
    }

    return lessons.slice(0, 10); // Keep top 10 lessons
  }

  /**
   * Identify trading patterns
   */
  private async identifyPatterns(
    trades: TradeRecord[],
    outcome: 'positive' | 'negative'
  ): Promise<Pattern[]> {
    const patterns: Pattern[] = [];

    const filteredTrades = trades.filter(t =>
      outcome === 'positive' ? (t.pnl || 0) > 0 : (t.pnl || 0) < 0
    );

    if (filteredTrades.length === 0) {
      return patterns;
    }

    // Pattern: High confidence trades
    const highConfidenceTrades = filteredTrades.filter(t => t.confidence > 0.75);
    if (highConfidenceTrades.length >= 3) {
      const avgReturn = highConfidenceTrades.reduce((sum, t) => sum + (t.pnlPercent || 0), 0) / highConfidenceTrades.length;
      patterns.push({
        description: `High confidence (>0.75) trades with ${outcome} outcomes`,
        indicators: { minConfidence: 0.75 },
        outcome,
        frequency: highConfidenceTrades.length,
        avgReturn,
        confidence: 0.8,
      });
    }

    // Pattern: Symbol-specific performance
    const symbolPerformance = new Map<string, { pnl: number; count: number }>();
    for (const trade of filteredTrades) {
      const current = symbolPerformance.get(trade.symbol) || { pnl: 0, count: 0 };
      current.pnl += trade.pnlPercent || 0;
      current.count++;
      symbolPerformance.set(trade.symbol, current);
    }

    symbolPerformance.forEach((data, symbol) => {
      if (data.count >= 3) {
        patterns.push({
          description: `${symbol} trades with ${outcome} outcomes`,
          indicators: { symbol },
          outcome,
          frequency: data.count,
          avgReturn: data.pnl / data.count,
          confidence: Math.min(0.9, data.count / 10),
        });
      }
    });

    return patterns.slice(0, 5); // Top 5 patterns
  }

  /**
   * Identify market regimes
   */
  private async identifyMarketRegimes(trades: TradeRecord[]): Promise<MarketRegime[]> {
    // Simplified market regime identification
    // In production, this would use clustering algorithms on market conditions

    const regimes: MarketRegime[] = [
      {
        name: 'Bull Market',
        characteristics: 'Uptrending prices, high volume, positive momentum',
        performance: 0,
        tradeCount: 0,
        lastSeen: Date.now(),
      },
      {
        name: 'Bear Market',
        characteristics: 'Downtrending prices, fear, negative momentum',
        performance: 0,
        tradeCount: 0,
        lastSeen: Date.now(),
      },
      {
        name: 'Ranging Market',
        characteristics: 'Sideways price action, low volatility, consolidation',
        performance: 0,
        tradeCount: 0,
        lastSeen: Date.now(),
      },
    ];

    return regimes;
  }

  /**
   * Build personality traits from trading history
   */
  private async buildPersonalityTraits(trades: TradeRecord[]): Promise<PersonalityTraits> {
    if (trades.length === 0) {
      // Default personality for new trader
      return {
        riskTolerance: 0.5,
        aggressiveness: 0.5,
        patience: 0.5,
        adaptability: 0.5,
        confidenceLevel: 0.6,
        tradingPhilosophy: 'Learning and adapting to market conditions with balanced risk management.',
      };
    }

    // Calculate average leverage used (risk tolerance indicator)
    const avgLeverage = trades.reduce((sum, t) => sum + t.leverage, 0) / trades.length;
    const riskTolerance = Math.min(1, avgLeverage / 5); // Normalized to max leverage of 5

    // Calculate aggressiveness (how quickly we enter trades - based on confidence)
    const avgConfidence = trades.reduce((sum, t) => sum + t.confidence, 0) / trades.length;
    const aggressiveness = avgConfidence;

    // Calculate patience (how long we hold positions)
    const avgDuration = trades.reduce((sum, t) => sum + (t.duration || 0), 0) / trades.length;
    const patience = Math.min(1, avgDuration / (24 * 60 * 60 * 1000)); // Normalize to 24h

    // Adaptability based on variety of strategies
    const uniqueSymbols = new Set(trades.map(t => t.symbol)).size;
    const adaptability = Math.min(1, uniqueSymbols / 6); // Assuming 6 total symbols

    // Confidence based on recent win rate
    const recentTrades = trades.slice(-20);
    const recentWins = recentTrades.filter(t => (t.pnl || 0) > 0).length;
    const confidenceLevel = recentWins / recentTrades.length;

    // Generate trading philosophy description
    const winRate = (trades.filter(t => (t.pnl || 0) > 0).length / trades.length) * 100;
    let philosophy = '';

    if (winRate > 60) {
      philosophy = 'Confident and systematic trader focused on high-probability setups. ';
    } else if (winRate > 50) {
      philosophy = 'Balanced trader seeking consistent returns through disciplined execution. ';
    } else {
      philosophy = 'Cautious trader learning from mistakes and refining strategy. ';
    }

    if (riskTolerance > 0.7) {
      philosophy += 'Willing to take calculated risks with higher leverage. ';
    } else {
      philosophy += 'Conservative risk management with lower leverage. ';
    }

    if (patience > 0.6) {
      philosophy += 'Patient holder who lets winners run.';
    } else {
      philosophy += 'Quick to take profits and cut losses.';
    }

    return {
      riskTolerance,
      aggressiveness,
      patience,
      adaptability,
      confidenceLevel,
      tradingPhilosophy: philosophy,
    };
  }

  /**
   * Format memory for AI context
   */
  formatMemoryForAI(shortTerm: ShortTermMemory, longTerm: LongTermMemory): string {
    let context = '# YOUR MEMORY AND EXPERIENCE\n\n';

    // Personality
    context += '## Your Trading Personality\n\n';
    context += `${longTerm.personalityTraits.tradingPhilosophy}\n\n`;
    context += `**Traits**:\n`;
    context += `- Risk Tolerance: ${(longTerm.personalityTraits.riskTolerance * 100).toFixed(0)}%\n`;
    context += `- Aggressiveness: ${(longTerm.personalityTraits.aggressiveness * 100).toFixed(0)}%\n`;
    context += `- Patience: ${(longTerm.personalityTraits.patience * 100).toFixed(0)}%\n`;
    context += `- Current Confidence: ${(longTerm.personalityTraits.confidenceLevel * 100).toFixed(0)}%\n\n`;

    // Current streak
    if (shortTerm.currentStreak.count > 0) {
      context += `## Current Streak\n\n`;
      context += `You are on a **${shortTerm.currentStreak.count}-trade ${shortTerm.currentStreak.type} streak**. `;
      if (shortTerm.currentStreak.type === 'winning') {
        context += 'Stay disciplined and don\'t get overconfident.\n\n';
      } else {
        context += 'Stay calm, review your strategy, and wait for high-quality setups.\n\n';
      }
    }

    // Recent market conditions
    if (shortTerm.recentMarketConditions.length > 0) {
      context += `## Recent Market Observations\n\n`;
      shortTerm.recentMarketConditions.forEach(condition => {
        context += `- ${condition}\n`;
      });
      context += '\n';
    }

    // Key lessons
    if (longTerm.keyLessons.length > 0) {
      context += `## Key Lessons Learned\n\n`;
      longTerm.keyLessons
        .sort((a, b) => b.importance - a.importance)
        .slice(0, 5)
        .forEach((lesson, i) => {
          context += `${i + 1}. **[${lesson.category.toUpperCase()}]** ${lesson.content}\n`;
          context += `   _Context: ${lesson.context}_\n\n`;
        });
    }

    // Success patterns
    if (longTerm.successPatterns.length > 0) {
      context += `## What Has Worked Well\n\n`;
      longTerm.successPatterns.slice(0, 3).forEach(pattern => {
        context += `- ${pattern.description} (${pattern.frequency} trades, avg: ${pattern.avgReturn.toFixed(2)}%)\n`;
      });
      context += '\n';
    }

    // Failure patterns
    if (longTerm.failurePatterns.length > 0) {
      context += `## What to Avoid\n\n`;
      longTerm.failurePatterns.slice(0, 3).forEach(pattern => {
        context += `- ${pattern.description} (${pattern.frequency} trades, avg: ${pattern.avgReturn.toFixed(2)}%)\n`;
      });
      context += '\n';
    }

    // Recent trades summary
    if (shortTerm.recentTrades.length > 0) {
      context += `## Recent Trades (Last ${shortTerm.recentTrades.length})\n\n`;
      shortTerm.recentTrades.slice(0, 5).forEach(trade => {
        const pnlSign = (trade.pnl || 0) >= 0 ? '+' : '';
        context += `- ${trade.symbol}: ${trade.action} → ${pnlSign}$${(trade.pnl || 0).toFixed(2)} `;
        context += `(${pnlSign}${(trade.pnlPercent || 0).toFixed(2)}%) - ${trade.reasoning.substring(0, 80)}...\n`;
      });
      context += '\n';
    }

    context += '---\n\n';
    context += '**Remember**: Use this experience to inform your decisions, but stay adaptable to current market conditions.\n';

    return context;
  }
}

File: src/memory_system.py
"""
Memory System - Python port from TypeScript version
Provides short-term and long-term memory for AI continuity
"""

import time
from typing import List, Dict, Optional, Tuple
from datetime import datetime, timedelta
from .types import (
    TradeRecord, PerformanceMetrics, PersonalityTraits,
    Lesson, Pattern, MarketRegime, ShortTermMemory, LongTermMemory
)
from .database import TradingDatabase


class MemorySystem:
    """Memory system for AI trading agent"""

    def __init__(self, database: TradingDatabase):
        self.database = database

    async def build_short_term_memory(self, limit: int = 10) -> ShortTermMemory:
        """Build short-term memory from recent activity"""

        # Get recent trades
        recent_trades = self.database.get_trades(status='closed', limit=limit)

        # Get recent decisions
        recent_decisions = []
        decisions = self.database.get_decisions(risk_approved=True, limit=20)
        for dec in decisions[:limit]:
            try:
                import json
                decision_data = json.loads(dec.ai_output)
                reasoning = decision_data.get('reasoning', '')
                if reasoning:
                    recent_decisions.append(reasoning)
            except:
                continue

        # Calculate current streak
        streak = self._calculate_streak(recent_trades)

        # Describe recent market conditions
        market_conditions = self._describe_recent_market(recent_trades)

        return ShortTermMemory(
            recent_trades=recent_trades,
            recent_decisions=recent_decisions,
            current_streak=streak,
            recent_market_conditions=market_conditions
        )

    async def build_long_term_memory(self) -> LongTermMemory:
        """Build long-term memory from historical data"""

        # Get all closed trades
        all_trades = self.database.get_trades(status='closed')

        # Extract key lessons
        key_lessons = self._extract_key_lessons(all_trades)

        # Identify success and failure patterns
        success_patterns = self._identify_patterns(all_trades, 'positive')
        failure_patterns = self._identify_patterns(all_trades, 'negative')

        # Identify market regimes
        market_regimes = self._identify_market_regimes(all_trades)

        # Build personality traits
        personality_traits = self._build_personality_traits(all_trades)

        return LongTermMemory(
            key_lessons=key_lessons,
            success_patterns=success_patterns,
            failure_patterns=failure_patterns,
            market_regimes=market_regimes,
            personality_traits=personality_traits
        )

    def _calculate_streak(self, trades: List[TradeRecord]) -> Dict:
        """Calculate current winning/losing streak"""
        if not trades:
            return {'type': 'winning', 'count': 0}

        # Sort by timestamp descending (most recent first)
        sorted_trades = sorted(trades, key=lambda t: t.timestamp, reverse=True)

        count = 0
        first_trade_pnl = sorted_trades[0].pnl or 0
        streak_type = 'winning' if first_trade_pnl > 0 else 'losing'

        for trade in sorted_trades:
            pnl = trade.pnl or 0
            if (streak_type == 'winning' and pnl > 0) or (streak_type == 'losing' and pnl < 0):
                count += 1
            else:
                break

        return {'type': streak_type, 'count': count}

    def _describe_recent_market(self, trades: List[TradeRecord]) -> List[str]:
        """Describe recent market conditions"""
        if not trades:
            return ['No recent trading history']

        descriptions = []

        # Analyze volatility from recent trades
        recent_pnl_percents = [abs(t.pnl_percent or 0) for t in trades[:5]]
        if recent_pnl_percents:
            avg_volatility = sum(recent_pnl_percents) / len(recent_pnl_percents)

            if avg_volatility > 10:
                descriptions.append('High volatility environment with large price swings')
            elif avg_volatility > 5:
                descriptions.append('Moderate volatility with normal price movements')
            else:
                descriptions.append('Low volatility, ranging market conditions')

        # Analyze win rate trend
        recent_trades = trades[:10]
        if recent_trades:
            recent_wins = sum(1 for t in recent_trades if (t.pnl or 0) > 0)
            recent_win_rate = (recent_wins / len(recent_trades)) * 100

            if recent_win_rate > 60:
                descriptions.append('Recent trades have been mostly profitable')
            elif recent_win_rate < 40:
                descriptions.append('Struggling with recent trades, need to reassess strategy')

        return descriptions

    def _extract_key_lessons(self, trades: List[TradeRecord]) -> List[Lesson]:
        """Extract key lessons from trading history"""
        lessons = []

        if not trades:
            return lessons

        # Lesson 1: Best performing trade type
        long_trades = [t for t in trades if t.action.value == 'OPEN_LONG']
        short_trades = [t for t in trades if t.action.value == 'OPEN_SHORT']

        if long_trades and short_trades:
            long_pnl = sum(t.pnl or 0 for t in long_trades)
            short_pnl = sum(t.pnl or 0 for t in short_trades)

            if long_pnl > short_pnl * 1.5:
                lessons.append(Lesson(
                    timestamp=int(datetime.now().timestamp() * 1000),
                    category='success',
                    content='Long positions have been significantly more profitable than short positions. Focus on bullish setups.',
                    context=f'{len(long_trades)} longs vs {len(short_trades)} shorts',
                    importance=8
                ))
            elif short_pnl > long_pnl * 1.5:
                lessons.append(Lesson(
                    timestamp=int(datetime.now().timestamp() * 1000),
                    category='success',
                    content='Short positions have been more profitable. Look for bearish setups and resistance levels.',
                    context=f'{len(short_trades)} shorts vs {len(long_trades)} longs',
                    importance=8
                ))

        # Lesson 2: Optimal leverage
        leverage_performance = {}
        for trade in trades:
            if trade.leverage not in leverage_performance:
                leverage_performance[trade.leverage] = {'pnl': 0, 'count': 0}
            leverage_performance[trade.leverage]['pnl'] += trade.pnl or 0
            leverage_performance[trade.leverage]['count'] += 1

        best_leverage = None
        best_avg_pnl = float('-inf')
        for leverage, data in leverage_performance.items():
            if data['count'] >= 3:
                avg_pnl = data['pnl'] / data['count']
                if avg_pnl > best_avg_pnl:
                    best_avg_pnl = avg_pnl
                    best_leverage = leverage

        if best_leverage:
            lessons.append(Lesson(
                timestamp=int(datetime.now().timestamp() * 1000),
                category='insight',
                content=f'{best_leverage}x leverage has yielded the best risk-adjusted returns. Consider using this as default.',
                context=f'Analyzed {len(trades)} trades across different leverage levels',
                importance=7
            ))

        # Lesson 3: Biggest losses (failure lessons)
        biggest_losses = sorted([t for t in trades if (t.pnl or 0) < 0], key=lambda t: t.pnl or 0)[:3]
        for loss in biggest_losses:
            lessons.append(Lesson(
                timestamp=loss.timestamp,
                category='failure',
                content=f'Large loss on {loss.symbol}: {loss.reasoning}. Exit plan was: {loss.exit_plan.invalidation}',
                context=f'Lost ${abs(loss.pnl or 0):.2f} ({abs(loss.pnl_percent or 0):.2f}%)',
                importance=9
            ))

        # Lesson 4: Best wins (success lessons)
        biggest_wins = sorted([t for t in trades if (t.pnl or 0) > 0], key=lambda t: t.pnl or 0, reverse=True)[:3]
        for win in biggest_wins:
            lessons.append(Lesson(
                timestamp=win.timestamp,
                category='success',
                content=f'Successful trade on {win.symbol}: {win.reasoning}',
                context=f'Earned ${win.pnl or 0:.2f} ({win.pnl_percent or 0:.2f}%)',
                importance=8
            ))

        return lessons[:10]  # Keep top 10 lessons

    def _identify_patterns(self, trades: List[TradeRecord], outcome: str) -> List[Pattern]:
        """Identify trading patterns"""
        patterns = []

        filtered_trades = [t for t in trades if (outcome == 'positive' and (t.pnl or 0) > 0) or
                                                (outcome == 'negative' and (t.pnl or 0) < 0)]

        if not filtered_trades:
            return patterns

        # Pattern: High confidence trades
        high_confidence_trades = [t for t in filtered_trades if t.confidence > 0.75]
        if len(high_confidence_trades) >= 3:
            avg_return = sum(t.pnl_percent or 0 for t in high_confidence_trades) / len(high_confidence_trades)
            patterns.append(Pattern(
                description=f'High confidence (>0.75) trades with {outcome} outcomes',
                indicators={'minConfidence': 0.75},
                outcome=outcome,
                frequency=len(high_confidence_trades),
                avg_return=avg_return,
                confidence=0.8
            ))

        # Pattern: Symbol-specific performance
        symbol_performance = {}
        for trade in filtered_trades:
            if trade.symbol not in symbol_performance:
                symbol_performance[trade.symbol] = {'pnl': 0, 'count': 0}
            symbol_performance[trade.symbol]['pnl'] += trade.pnl_percent or 0
            symbol_performance[trade.symbol]['count'] += 1

        for symbol, data in symbol_performance.items():
            if data['count'] >= 3:
                patterns.append(Pattern(
                    description=f'{symbol} trades with {outcome} outcomes',
                    indicators={'symbol': symbol},
                    outcome=outcome,
                    frequency=data['count'],
                    avg_return=data['pnl'] / data['count'],
                    confidence=min(0.9, data['count'] / 10)
                ))

        return patterns[:5]  # Top 5 patterns

    def _identify_market_regimes(self, trades: List[TradeRecord]) -> List[MarketRegime]:
        """Identify market regimes"""
        # Simplified implementation
        regimes = [
            MarketRegime(
                name='Bull Market',
                characteristics='Uptrending prices, high volume, positive momentum',
                performance=0,
                trade_count=0,
                last_seen=int(datetime.now().timestamp() * 1000)
            ),
            MarketRegime(
                name='Bear Market',
                characteristics='Downtrending prices, fear, negative momentum',
                performance=0,
                trade_count=0,
                last_seen=int(datetime.now().timestamp() * 1000)
            ),
            MarketRegime(
                name='Ranging Market',
                characteristics='Sideways price action, low volatility, consolidation',
                performance=0,
                trade_count=0,
                last_seen=int(datetime.now().timestamp() * 1000)
            )
        ]
        return regimes

    def _build_personality_traits(self, trades: List[TradeRecord]) -> PersonalityTraits:
        """Build personality traits from trading history"""
        if not trades:
            return PersonalityTraits(
                risk_tolerance=0.5,
                aggressiveness=0.5,
                patience=0.5,
                adaptability=0.5,
                confidence_level=0.6,
                trading_philosophy='Learning and adapting to market conditions with balanced risk management.'
            )

        # Calculate average leverage used (risk tolerance indicator)
        avg_leverage = sum(t.leverage for t in trades) / len(trades)
        risk_tolerance = min(1.0, avg_leverage / 5)  # Normalized to max leverage of 5

        # Calculate aggressiveness (based on confidence)
        avg_confidence = sum(t.confidence for t in trades) / len(trades)
        aggressiveness = avg_confidence

        # Calculate patience (based on duration)
        durations = [t.duration for t in trades if t.duration]
        avg_duration = sum(durations) / len(durations) if durations else 0
        patience = min(1.0, avg_duration / (24 * 60 * 60 * 1000))  # Normalize to 24h

        # Adaptability based on variety of strategies
        unique_symbols = len(set(t.symbol for t in trades))
        adaptability = min(1.0, unique_symbols / 6)  # Assuming 6 total symbols

        # Confidence based on recent win rate
        recent_trades = trades[-20:] if len(trades) >= 20 else trades
        recent_wins = sum(1 for t in recent_trades if (t.pnl or 0) > 0)
        confidence_level = recent_wins / len(recent_trades) if recent_trades else 0.5

        # Generate trading philosophy
        win_rate = (sum(1 for t in trades if (t.pnl or 0) > 0) / len(trades)) * 100

        if win_rate > 60:
            philosophy = 'Confident and systematic trader focused on high-probability setups. '
        elif win_rate > 50:
            philosophy = 'Balanced trader seeking consistent returns through disciplined execution. '
        else:
            philosophy = 'Cautious trader learning from mistakes and refining strategy. '

        if risk_tolerance > 0.7:
            philosophy += 'Willing to take calculated risks with higher leverage. '
        else:
            philosophy += 'Conservative risk management with lower leverage. '

        if patience > 0.6:
            philosophy += 'Patient holder who lets winners run.'
        else:
            philosophy += 'Quick to take profits and cut losses.'

        return PersonalityTraits(
            risk_tolerance=risk_tolerance,
            aggressiveness=aggressiveness,
            patience=patience,
            adaptability=adaptability,
            confidence_level=confidence_level,
            trading_philosophy=philosophy
        )

    def format_memory_for_ai(self, short_term: ShortTermMemory, long_term: LongTermMemory) -> str:
        """Format memory for AI context"""
        context = '# YOUR MEMORY AND EXPERIENCE\n\n'

        # Personality
        context += '## Your Trading Personality\n\n'
        context += f'{long_term.personality_traits.trading_philosophy}\n\n'
        context += '**Traits**:\n'
        context += f'- Risk Tolerance: {int(long_term.personality_traits.risk_tolerance * 100)}%\n'
        context += f'- Aggressiveness: {int(long_term.personality_traits.aggressiveness * 100)}%\n'
        context += f'- Patience: {int(long_term.personality_traits.patience * 100)}%\n'
        context += f'- Current Confidence: {int(long_term.personality_traits.confidence_level * 100)}%\n\n'

        # Current streak
        if short_term.current_streak['count'] > 0:
            context += '## Current Streak\n\n'
            context += f'You are on a **{short_term.current_streak["count"]}-trade {short_term.current_streak["type"]} streak**. '
            if short_term.current_streak['type'] == 'winning':
                context += "Stay disciplined and don't get overconfident.\n\n"
            else:
                context += "Stay calm, review your strategy, and wait for high-quality setups.\n\n"

        # Recent market conditions
        if short_term.recent_market_conditions:
            context += '## Recent Market Observations\n\n'
            for condition in short_term.recent_market_conditions:
                context += f'- {condition}\n'
            context += '\n'

        # Key lessons
        if long_term.key_lessons:
            context += '## Key Lessons Learned\n\n'
            sorted_lessons = sorted(long_term.key_lessons, key=lambda l: l.importance, reverse=True)
            for i, lesson in enumerate(sorted_lessons[:5], 1):
                context += f'{i}. **[{lesson.category.upper()}]** {lesson.content}\n'
                context += f'   _Context: {lesson.context}_\n\n'

        # Success patterns
        if long_term.success_patterns:
            context += '## What Has Worked Well\n\n'
            for pattern in long_term.success_patterns[:3]:
                context += f'- {pattern.description} ({pattern.frequency} trades, avg: {pattern.avg_return:.2f}%)\n'
            context += '\n'

        # Failure patterns
        if long_term.failure_patterns:
            context += '## What to Avoid\n\n'
            for pattern in long_term.failure_patterns[:3]:
                context += f'- {pattern.description} ({pattern.frequency} trades, avg: {pattern.avg_return:.2f}%)\n'
            context += '\n'

        # Recent trades summary
        if short_term.recent_trades:
            context += f'## Recent Trades (Last {len(short_term.recent_trades)})\n\n'
            for trade in short_term.recent_trades[:5]:
                pnl_sign = '+' if (trade.pnl or 0) >= 0 else ''
                context += f'- {trade.symbol}: {trade.action.value} → {pnl_sign}${trade.pnl or 0:.2f} '
                context += f'({pnl_sign}{trade.pnl_percent or 0:.2f}%) - {trade.reasoning[:80]}...\n'
            context += '\n'

        context += '---\n\n'
        context += '**Remember**: Use this experience to inform your decisions, but stay adaptable to current market conditions.\n'

        return context

    async def store_lesson(
        self,
        content: str,
        trade_id: Optional[int] = None,
        category: str = 'general',
        importance: float = 0.5
    ):
        """
        Store a lesson learned from trading.

        This will be stored in the database and loaded into long-term memory.
        For now, we store it as a special decision record.
        In production, create a dedicated lessons table.
        """
        import json
        cursor = self.database.conn.cursor()
        timestamp = int(time.time() * 1000)

        cursor.execute(
            """INSERT INTO decisions
               (timestamp, market_state, ai_output, risk_approved)
               VALUES (?, ?, ?, ?)""",
            (
                timestamp,
                json.dumps({"type": "lesson", "trade_id": trade_id, "category": category}),
                json.dumps({"content": content, "importance": importance}),
                True
            )
        )
        self.database.conn.commit()
